lib/active_merchant/billing/gateways/merchant_partners.rb

```diff
@@ -0,0 +1,253 @@
+
+        commit("refund", post)
+      end
+
+      def credit(amount, payment_method, options={})
+        post = {}
+        add_invoice(post, amount, options)
+        add_payment_method(post, payment_method)
+
+        commit(payment_method.is_a?(String) ? "stored_credit" : "credit", post)
+      end
+
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def store(payment_method, options = {})
+        post = {}
+        add_payment_method(post, payment_method)
+        add_customer_data(post, options)
+
+        # 3 == no tx, just store
+        post[:profileactiontype] = options[:profileactiontype] || "3"
+
+        commit("store", post)
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((<ccnum>)[^<]+(<))i, '\1[FILTERED]\2').
+          gsub(%r((<cvv2>)[^<]+(<))i, '\1[FILTERED]\2').
+          gsub(%r((<merchantpin>)[^<]+(<))i, '\1[FILTERED]\2')
+      end
+
+      def test?
+        @options[:account_id].eql?('TEST0')
+      end
+
+      private
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money)
+        post[:merchantordernumber] = options[:order_id]
+        post[:currency] = options[:currency] || currency(money)
+      end
+
+      def add_payment_method(post, payment_method)
+        if(payment_method.is_a?(String))
+          user_profile_id, last_4 = split_authorization(payment_method)
+          post[:userprofileid] = user_profile_id
+          post[:last4digits] = last_4
+        else
+          post[:ccname] = payment_method.name
+          post[:ccnum] = payment_method.number
+          post[:cvv2] = payment_method.verification_value
+          post[:expmon] = format(payment_method.month, :two_digits)
+          post[:expyear] = format(payment_method.year, :four_digits)
+          post[:swipedata] = payment_method.track_data if payment_method.track_data
+        end
+      end
+
+      def add_customer_data(post, options)
+        post[:email] = options[:email] if options[:email]
+        post[:ipaddress] = options[:ip] if options[:ip]
+        if(billing_address = (options[:billing_address] || options[:address]))
+          post[:billaddr1] = billing_address[:address1]
+          post[:billaddr2] = billing_address[:address2]
+          post[:billcity] = billing_address[:city]
+          post[:billstate] = billing_address[:state]
+          post[:billcountry] = billing_address[:country]
+          post[:bilzip]    = billing_address[:zip]
+          post[:phone] = billing_address[:phone]
+        end
+      end
+
+      def add_reference(post, authorization)
+        post[:historykeyid] = authorization
+      end
+
+      ACTIONS = {
+        "purchase" => "2",
```

Comment:
I still prefer symbols for the keys of such maps, but I know not all devs agree with me.  Down with the hashrocket!